### PR TITLE
Update menu item

### PR DESCRIPTION
**Pull request by CPerinet**
- Add support for a `small` size on MenuItem & MenuItemV2
- On MenuItemV2, remove the constrain that prevents showing both a tick and and icon when `roleStructure=”listoption”`
- Add a MenuSection component that productize what is currently done using MenuTitle and MenuDividers. It can also be extended to include logic for rendering ticked MenuItems
- `selected=true` no longer extend the styles of `active=true`

### CODE REVIEW

File: packages/core/src/common/classes.ts [hunk 1]
```diff
@@ -193,7 +193,12 @@ export const FORM_HELPER_TEXT = `${NS}-form-helper-text`;
 export const FORM_GROUP_SUB_LABEL = `${NS}-form-group-sub-label`;
 
 export const MENU = `${NS}-menu`;
+export const MENU_NO_PADDING = `${MENU}-no-padding`;
```
Comment (CPerinet) [inline packages/core/src/common/classes.ts:196]: Used to allow MenuSection to handle their own padding thus having a divider that goes edge to edge.

File: packages/core/src/common/classes.ts [hunk 2]
```diff
@@ -193,7 +193,12 @@ export const FORM_HELPER_TEXT = `${NS}-form-helper-text`;
...
+export const MENU_SECTION = `${MENU}-section`;
 export const MENU_ITEM = `${MENU}-item`;
+export const MENU_ITEM_INDENT = `${MENU_ITEM}-indent`;
```
Comment (CPerinet) [inline packages/core/src/common/classes.ts:199]: Applied when `roleStructure=listoption` and `selected=false`. This creates a gap where the tick should be so the text lines up horizontally. Should only be used when one item is set to `selected=true` but not enforced.

File: packages/core/src/components/menu/_common.scss [hunk 3]
```diff
@@ -22,12 +22,16 @@ $menu-item-padding-large: ($pt-button-height-large - $pt-icon-size-large) * 0.5
 $menu-item-padding-vertical: ($pt-button-height - $menu-item-line-height) * 0.5 !default;
 $menu-item-padding-vertical-large:
   ($pt-button-height-large - $menu-item-line-height-large) * 0.5 !default;
+$menu-item-padding-small: ($pt-button-height-small - $pt-icon-size-standard) * 0.5 !default;
```
Comment (CPerinet) [inline packages/core/src/components/menu/_common.scss:25]: Adding padding variables for the small variant.

File: packages/core/src/components/menu/_common.scss [hunk 4]
```diff
@@ -22,12 +22,16 @@ $menu-item-padding-large: ($pt-button-height-large - $pt-icon-size-large) * 0.5
...
+$menu-item-padding-vertical-small:
+  ($pt-button-height-small - $menu-item-line-height) * 0.5 !default;
+$menu-item-selected-icon-spacing: 2px;
```
Comment (CPerinet) [inline packages/core/src/components/menu/_common.scss:28]: Variable for the tick icon spacing. Used for setting the spacing and computing the indent size below.

File: packages/core/src/components/menu/_common.scss [hunk 5]
```diff
@@ -22,12 +22,16 @@ $menu-item-padding-large: ($pt-button-height-large - $pt-icon-size-large) * 0.5
...
+$menu-item-indent: $pt-icon-size-standard + $menu-item-selected-icon-spacing * 2;
```
Comment (CPerinet) [inline packages/core/src/components/menu/_common.scss:29]: Size of the indent

File: packages/core/src/components/menu/_common.scss [hunk 6]
```diff
@@ -70,6 +74,7 @@ $dark-menu-item-intent-colors: (
 
   &::before,
   .#{$ns}-menu-item-icon,
+  .#{$ns}-menu-item-tick,
```
Comment (CPerinet) [inline packages/core/src/components/menu/_common.scss:78]: menu-item-tick should have the same color as menu-icon. Adding the selector everywhere item-icon color is set.

File: packages/core/src/components/menu/_common.scss [hunk 7]
```diff
@@ -85,20 +90,26 @@ $dark-menu-item-intent-colors: (
 
   &:active,
   &.#{$ns}-active {
-    background-color: $menu-item-color-active;
+    @include menu-item-active(false);
```
Comment (CPerinet) [inline packages/core/src/components/menu/_common.scss:94]: Active state used to have this default value but it was always overridden by the selected modifier that I removed below. Including the mixin keep the state where no intent active item is rendered blue.

File: packages/core/src/components/menu/_common.scss [hunk 8]
```diff
@@ -85,20 +90,26 @@ $dark-menu-item-intent-colors: (
...
+  &.#{$ns}-menu-item-ticked {
+    padding-left: 0 !important;
+
+    .#{$ns}-menu-item-tick {
+      margin: 0 $menu-item-selected-icon-spacing !important;
     }
   }
+  /* stylelint-enable declaration-no-important */
 
-  &.#{$ns}-selected {
```
Comment (CPerinet) [inline packages/core/src/components/menu/_common.scss:95]: De-extending active styles to the selected prop.

File: packages/core/src/components/menu/_common.scss [hunk 9]
```diff
@@ -85,20 +90,26 @@ $dark-menu-item-intent-colors: (
...
-    &,
-    &:hover,
-    &:active {
-      @include menu-item-selected-active(false);
-    }
+  // pt-disable always overrides over styles
+  /* stylelint-disable declaration-no-important */
+  &.#{$ns}-menu-item-indent {
```
Comment (CPerinet) [inline packages/core/src/components/menu/_common.scss:110]: The small and large modifier wins priority if no important is set here.

File: packages/core/src/components/menu/_common.scss [hunk 10]
```diff
@@ -146,14 +158,6 @@ $dark-menu-item-intent-colors: (
     }
   }
 
-  &.#{$ns}-selected {
```
Comment (CPerinet) [inline packages/core/src/components/menu/_common.scss:149]: De-extending active styles to the selected prop.

File: packages/core/src/components/menu/_common.scss [hunk 11]
```diff
@@ -176,11 +180,7 @@ $dark-menu-item-intent-colors: (
   background-color: $menu-item-color-hover;
   color: inherit;
   cursor: pointer;
-  text-decoration: none;
 
-  &.#{ns}-selected {
```
Comment (CPerinet) [inline packages/core/src/components/menu/_common.scss:181]: De-extending active styles to the selected prop.

File: packages/core/src/components/menu/_common.scss [hunk 12]
```diff
@@ -191,13 +191,9 @@ $dark-menu-item-intent-colors: (
   .#{$ns}-submenu-icon {
     color: $pt-dark-icon-color;
   }
-
-  &.#{ns}-selected {
```
Comment (CPerinet) [inline packages/core/src/components/menu/_common.scss:195]: De-extending active styles to the selected prop.

File: packages/core/src/components/menu/_common.scss [hunk 13]
```diff
@@ -191,13 +191,9 @@ $dark-menu-item-intent-colors: (
...
-    @include menu-item-selected-active(true);
-  }
 }
 
-@mixin menu-item-selected-active($is-dark) {
+@mixin menu-item-active($is-dark) {
```
Comment (CPerinet) [inline packages/core/src/components/menu/_common.scss:197]: Rename to reflect the de-extending active styles to the selected prop.

File: packages/core/src/components/menu/menuSection.tsx [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+import { Props } from "../../common/props";
+import { MenuItem, MenuItemProps } from "./menuItem";
+import { H6 } from "../html/html";
+
+// eslint-disable-next-line deprecation/deprecation
+export type MenuSectionProps = IMenuSectionProps;
+
+export type MenuItemPropsWithId = MenuItemProps & {
+    // Unique identifier 
+    id: string;
```
Comment (CPerinet) [inline packages/core/src/components/menu/menuSection.tsx:29]: Needed for keys

File: packages/docs-app/src/examples/core-examples/menuItemExample.tsx [hunk 15]
```diff
@@ -20,35 +20,37 @@ import { Classes, H5, Intent, Menu, MenuItem, Switch } from "@blueprintjs/core";
 import { Example, ExampleProps, handleBooleanChange } from "@blueprintjs/docs-theme";
 
 import { IntentSelect } from "./common/intentSelect";
+import { Size, SizeSelect } from "./common/sizeSelect";
 
 export function MenuItemExample(props: ExampleProps) {
-    const [large, setLarge] = React.useState(false);
+    const [size, setSize] = React.useState<Size>("regular");
```
Comment (CPerinet) [inline packages/docs-app/src/examples/core-examples/menuItemExample.tsx:26]: Replace large selector by size selector

File: packages/docs-app/src/examples/popover2-examples/menuItem2Example.tsx [hunk 16]
```diff
@@ -48,7 +50,7 @@ export function MenuItem2Example(props: ExampleProps) {
                     isSelectedOptionAvailable ? undefined : (
                         <>
                             <Code>selected</Code> prop has no effect when <br />
-                            <Code>roleStructure="menuitem"</Code> or when an icon is set
```
Comment (CPerinet) [inline packages/docs-app/src/examples/popover2-examples/menuItem2Example.tsx:51]: Removing this constraint

**Comment by adidahiya** [comment]
Superseded by https://github.com/palantir/blueprint/pull/6269, let's continue discussion there. I merged develop into your branch. Now there is only one MenuItem, no more MenuItem2.

**Pull request closed without merging**